command/certificate/create.go

```diff
@@ -257,199 +341,306 @@ func createAction(ctx *cli.Context) error {
-				return errs.RequiredWithFlagValue(ctx, "profile", "self-signed", "subtle")
-			}
-			profile, err = x509util.NewSelfSignedLeafProfile(subject,
-				x509util.GenerateKeyPair(kty, crv, size),
-				x509util.WithNotBeforeAfterDuration(notBefore, notAfter, 0),
-				x509util.WithSANs(sans))
-			if err != nil {
-				return errors.WithStack(err)
-			}
-		default:
-			return errs.InvalidFlagValue(ctx, "profile", prof, "leaf, intermediate-ca, root-ca, self-signed")
+
+		ui.Printf("Your certificate signing request has been saved in %s.\n", crtFile)
+		if keyFile != "" {
+			ui.Printf("Your private key has been saved in %s.\n", keyFile)
 		}
-		var crtBytes []byte
-		crtBytes, err = profile.CreateCertificate()
+
+		return nil
+	}
+
+	// Bundle is only valid for leaf certificates
+	if bundle && profile != profileLeaf {
+		return errs.IncompatibleFlagValue(ctx, "bundle", "profile", profile)
+	}
+
+	// Subtle is required on self-signed certificates
+	if !subtle && profile == profileSelfSigned {
+		return errs.RequiredWithFlagValue(ctx, "profile", "self-signed", "subtle")
+	}
+
+	// Parse --ca and --ca-key flags and check when those flags are required.
+	parent, signer, err := parseSigner(ctx, priv)
+	if err != nil {
+		return err
+	}
+
+	var template string
+	var defaultValidity time.Duration
+	if templateFile != "" {
+		b, err := utils.ReadFile(templateFile)
 		if err != nil {
-			return errors.WithStack(err)
+			return err
 		}
-		pubPEMs = []*pem.Block{{
-			Type:  "CERTIFICATE",
-			Bytes: crtBytes,
-		}}
-		if bundle {
-			pubPEMs = append(pubPEMs, &pem.Block{
-				Type:  "CERTIFICATE",
-				Bytes: profile.Issuer().Raw,
-			})
+		template = string(b)
+		defaultValidity = defaultTemplatevalidity
+	} else {
+		// Use subject as default san for leaf or self-signed certificates
+		if len(sans) == 0 && (profile == profileLeaf || profile == profileSelfSigned) {
+			sans = append(sans, subject)
+		}
+
+		switch profile {
+		case profileLeaf:
+			template = x509util.DefaultLeafTemplate
+			defaultValidity = defaultLeafValidity
+		case profileIntermediateCA:
+			template = x509util.DefaultIntermediateTemplate
+			defaultValidity = defaultIntermediateValidity
+		case profileRootCA:
+			template = x509util.DefaultRootTemplate
+			defaultValidity = defaultRootValidity
+		case profileSelfSigned:
+			template = x509util.DefaultLeafTemplate
+			defaultValidity = defaultSelfSignedValidity
+		default:
+			return errs.InvalidFlagValue(ctx, "profile", profile, "leaf, intermediate-ca, root-ca, self-signed")
+		}
+	}
+
+	// Create X.509 certificate used as base for the certificate
+	cr, err := x509util.CreateCertificateRequest(subject, sans, signer)
+	if err != nil {
+		return err
+	}
+
+	// Create X.509 certificate
+	templateData := x509util.CreateTemplateData(subject, sans)
+	certificate, err := x509util.NewCertificate(cr, x509util.WithTemplate(template, templateData))
+	if err != nil {
+		return err
+	}
+	certTemplate := certificate.GetCertificate()
```

Comment:
ugh, these names are kinda confusing. GetCertificate() in my mind would return a Certificate. But in this case it's not actually signed, so it's really just returning a certificatetemplate or some other name for that. It's kinda confusing in practice. Maybe something we can update with 0.15.1

Reply:
sure